PARQUET-315: Add PARQUET_1_0 and non-repeated data performance tests …

spena opened the pull request:
Here's a list of changes I did on the parquet-benchmarks module:
- Add variable to test PARQUET_1_0 format
- Add variable to generate random data for the tests
- Move part of the DataGenerator.generate() code to constructors so that we measure time when
  when writing data only.
- Add annotation to display AverageTime. The Throughput mode was always displaying less than 1 op/s. So I thought time was a better mode here.

@nezihyigitbasi @danielcweeks You did a good job starting the parquet-benchmarks module. Could you please review these changes?

nezihyigitbasi commented on parquet-benchmarks/read-benchmark.sh:
```diff
@@ -0,0 +1,43 @@
+echo "Data generated, starting READ benchmarks"
+java -jar ${SCRIPT_PATH}/target/parquet-benchmarks.jar p*Read* "$@"
+echo "Cleaning up generated data"
+java -cp ${SCRIPT_PATH}/target/parquet-benchmarks.jar org.apache.parquet.benchmarks.DataGenerator cleanup
```
please add a newline here

nezihyigitbasi commented on parquet-benchmarks/run.sh:
```diff
@@ -19,13 +19,9 @@
-java -cp ${SCRIPT_PATH}/target/parquet-benchmarks.jar parquet.benchmarks.DataGenerator cleanup
\ No newline at end of file
+bash write-benchmark.sh "$@"
+bash read-benchmark.sh "$@"
```
newline

nezihyigitbasi commented on parquet-benchmarks/src/main/java/org/apache/parquet/benchmarks/ReadBenchmarks.java:
```diff
@@ -18,23 +18,36 @@
+import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.fs.Path;
-import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.*;
```
can you specify specific imports instead of wildcards?

nezihyigitbasi commented on parquet-benchmarks/src/main/java/org/apache/parquet/benchmarks/ReadBenchmarks.java:
```diff
@@ -18,23 +18,36 @@
 import java.io.IOException;
 
+@State(Scope.Benchmark)
+@BenchmarkMode(Mode.AverageTime)
```
why only average time? I think we can do `Mode.ALL` to get as many metrics as possible, which may provide useful info.

nezihyigitbasi commented:
@spena thanks for taking the time to create this PR. Can you also update the README?

nezihyigitbasi commented on parquet-benchmarks/pom.xml:
```diff
@@ -32,7 +32,7 @@
 
   <properties>
-    <jmh.version>1.3.4</jmh.version>
+    <jmh.version>1.9.3</jmh.version>
```
Is there a reason that you updated to `1.9.3` instead of the latest `1.10` release?

nezihyigitbasi commented on parquet-benchmarks/src/main/java/org/apache/parquet/benchmarks/DataGenerator.java:
```diff
@@ -126,14 +160,36 @@ public void cleanup()
+        System.err.println("  VERSION         Use a specific parquet file version to generate data.");
+        System.err.println("                  - v1 for PARQUET_1_0 version.");
+        System.err.println("                  - v2 for PARQUET_2_0 version.");
+        System.err.println("  -randomData     This flag specifies if random data will be used to generate data.");
```
Is `randomData` optional? When I only specify the version to generate some data (with a command that looks like ```.... org.apache.parquet.benchmarks.DataGenerator generate v2```) I get an exception

```Java
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 2
	at org.apache.parquet.benchmarks.DataGenerator.main(DataGenerator.java:188)
```

nezihyigitbasi commented on parquet-benchmarks/src/main/java/org/apache/parquet/benchmarks/DataGenerator.java:
```diff
@@ -126,14 +160,36 @@ public void cleanup()
+
+      String parquetVersion = args[1];
+
+      if (args.length == 3 && args[2].equalsIgnoreCase("-randomData")) {
```
Seems like to generate data I *have to* specify `-randomData`, then why is that a cmd line flag at all? I tried a few command line args to see how it behaves

- `java -cp ${SCRIPT_PATH}/target/parquet-benchmarks.jar org.apache.parquet.benchmarks.DataGenerator generate`

```
Generating test data ---> This shouldn't be printed btw.
Usage: generate VERSION [-randomData]

Options:
  VERSION         Use a specific parquet file version to generate data.
                  - v1 for PARQUET_1_0 version.
                  - v2 for PARQUET_2_0 version.
  -randomData     This flag specifies if random data will be used to generate data.
                  By default, only 1 random row is generated, and repeated across the file.
```

- `java -cp ${SCRIPT_PATH}/target/parquet-benchmarks.jar org.apache.parquet.benchmarks.DataGenerator generate v2`

```
Generating test data  ---> This shouldn't be printed btw.
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: 2
	at org.apache.parquet.benchmarks.DataGenerator.main(DataGenerator.java:188)
```
- `java -cp ${SCRIPT_PATH}/target/parquet-benchmarks.jar org.apache.parquet.benchmarks.DataGenerator generate v2 -randomData`

``` This one runs successfully ```

nezihyigitbasi commented:
@spena Added a few comments. Once you update the README and address the comments, I will be happy to give it a try. Thanks again!

nezihyigitbasi commented on parquet-benchmarks/write-benchmark.sh:
```diff
@@ -0,0 +1,28 @@
+SCRIPT_PATH=$( cd "$(dirname "$0")" ; pwd -P )
+
+echo "Starting WRITE benchmarks"
+java -jar ${SCRIPT_PATH}/target/parquet-benchmarks.jar p*Write* "$@"
```
newline

spena commented:
Thanks @nezihyigitbasi for the feedback.
I fixed the issues you mentioned. You will be able to generate data this time.

nezihyigitbasi commented on parquet-benchmarks/README.md:
````diff
@@ -25,10 +25,29 @@ First, build the ``parquet-benchmarks`` module
+
+```
+  # Run all benchmarks with PARQUET_2_0 and using random data
+  ./parquet-benchmarks/run.sh -wi 5 -i 5 -f 3 -p parquetVersion=v2 -p randomData=true
````
A user following the README will get the following error:

```
[/Users/nyigitbasi/other_workspace/parquet-mr] >> ./parquet-benchmarks/run.sh -wi 1 -i 1 -f 1 -p parquetVersion=v2 -p randomData=true
bash: write-benchmark.sh: No such file or directory
```

nezihyigitbasi commented on parquet-benchmarks/README.md:
````diff
@@ -25,10 +25,29 @@ First, build the ``parquet-benchmarks`` module
+Also, you can run only write or read benchmarks with the following command
+
+```
+ ./parquet-benchmarks/write-benchmarks.sh -wi 5 -i 5 -f 3 -p parquetVersion=v2
````
`read-benchmarks.sh` -> `read-benchmark.sh`
`write-benchmarks.sh` -> `write-benchmark.sh`

nezihyigitbasi commented on parquet-benchmarks/README.md:
```diff
@@ -25,10 +25,29 @@ First, build the ``parquet-benchmarks`` module
 mvn --projects parquet-benchmarks -amd -DskipTests -Denforcer.skip=true -P hadoop-2 clean package
```
I don't know what has changed in the mean time, but with the `hadoop-2` profile now the generated parquet-benchmarks jar doesn't contain `/META-INF/BenchmarkList` and it fails

```
[/Users/nyigitbasi/other_workspace/parquet-mr/parquet-benchmarks] >> ./run.sh -wi 1 -i 1 -f 1 -p parquetVersion=v2 -p randomData=true
Starting WRITE benchmarks
Exception in thread "main" java.lang.RuntimeException: ERROR: Unable to find the resource: /META-INF/BenchmarkList
	at org.openjdk.jmh.runner.AbstractResourceReader.getReaders(AbstractResourceReader.java:96)
	at org.openjdk.jmh.runner.BenchmarkList.find(BenchmarkList.java:104)
	at org.openjdk.jmh.runner.Runner.internalRun(Runner.java:228)
	at org.openjdk.jmh.runner.Runner.run(Runner.java:178)
	at org.openjdk.jmh.Main.main(Main.java:66)
```

A build without specifying any profile creates a valid jar though. So build command should be

```
mvn --projects parquet-benchmarks -amd -DskipTests -Denforcer.skip=true clean package
```

nezihyigitbasi commented on parquet-benchmarks/README.md:
```diff
@@ -25,10 +25,29 @@ First, build the ``parquet-benchmarks`` module
+and how well they compress data. But if you need more random data, then you can specify a flag so that files
+are tested with random data. By default, a sample of 100k random rows is generated on memory before writing them
+to disk.
+
```
Would be nice to rephrase this & clarify a little bit. How about something like:

```
While creating a test Parquet file of N records, by default all benchmarks use a single record 
that's repeated N times throughout the file. To have a larger number of random records in 
this test file you can specify `randomData=true` -- with this flag 100K random records 
are generated and used repeatedly while creating the test file. 
```

spena commented:
Thanks @nezihyigitbasi 
I added the changes based on your feedback. I liked your explanation about random data, so I replaced the old paragraph with yours.

I tried to figured out why the benchmark cannot be executed when compiling with hadoop-2, but I did not know why. So, I just replace the README to use the command with out the profile.

One thing. The read-benchmark.sh tests will fail when using parquetVersion=v2 and randomData=true due to the issue in PARQUET-152. If the patch is applied, then all tests run fine.

nezihyigitbasi commented:
@spena To keep the fixes to the existing benchmark module problems I created a new PR: https://github.com/apache/parquet-mr/pull/226

Yep I verified that your change makes the read-benchmark go through.

nezihyigitbasi commented on parquet-benchmarks/README.md:
````diff
@@ -22,13 +22,32 @@
+are generated and used repeatedly while creating the test file.
+
+```
+  # Run all benchmarks with PARQUET_2_0 and using random data
````
`... and using random data` --> `... and use random data`

nezihyigitbasi commented on parquet-benchmarks/README.md:
````diff
@@ -22,13 +22,32 @@
+  ./parquet-benchmarks/run.sh -wi 5 -i 5 -f 3 -p parquetVersion=v2     # Run all benchmarks with PARQUET_2_0
+```
+
+Also, you can run only write or read benchmarks with the following command
````
`command` --> `commands`

nezihyigitbasi commented on parquet-benchmarks/README.md:
````diff
@@ -22,13 +22,32 @@
 ```
 
-Then, you can run all the benchmarks with the following command
+Then, you can run all the benchmarks with the following command (use parquetVersion=vN to specify
````
`command` --> `commands`

nezihyigitbasi commented:
@spena I only added a few minor comments to the README. Other than that LGTM. 
@rdblue I have provided feedback to @spena and verified his changes. When he addresses my final comments this patch LGTM. It's all yours now.

spena commented:
Thanks @nezihyigitbasi.
I added the minor changes you mentioned.

Once #226 is committed, I will merge the code to my branch, and check it is applied correctly.

spena commented:
@nezihyigitbasi @rdblue 
I updated the branch with the latest changes of master (includes #226).

No more work to do here.

rdblue commented:
Thanks @spena, I'll take a look at this today.

julienledem commented:
@rdblue did you want to merge this?
@spena do you need to merge master
